### DISCUSSION TITLE
ピンの現在地から逆ジオコーディングしたい

### DISCUSSION THREAD

**Discussion by mm0o21** [lit-iphone-course/lit-iphone-course#56] [Question-original]
(1) 実現したいこと
ピンを刺した場合・刺さっているピンをタップした場合に出てくるハーフモーダルのラベルにそのピンの住所を表示したい
(2) 実現するために自分が試した内容とその結果
[https://dev.classmethod.jp/articles/geocoding-use-mapkit/](url)
この記事を参考に書いたのですが、viewcontrollerで取得した緯度経度の情報をどのように持っていけばわからない状態です。
(3) 調べた内容や資料
[https://qiita.com/PictoMki/items/bbda6df26291e40e89b8#%E3%83%94%E3%83%B3%E3%81%8C%E3%82%BF%E3%83%83%E3%83%97%E3%81%95%E3%82%8C%E3%81%9F%E6%99%82%E3%81%AE%E5%87%A6%E7%90%86](url)
[https://ymgsapo.com/2020/08/26/detect-pin-tapped-mkmap-view/](url)
[https://dev.classmethod.jp/articles/geocoding-use-mapkit/](url)
(4) 原因だと思われそうな箇所(ソースコードをコピペして貼る)
EditViewController
```swift
@IBAction func update(){
        //let coordinate =
        //let location = CLLocation(latitude://ピンの緯度, longitude://ピンの経度)
        
        CLGeocoder().reverseGeocodeLocation(location) { placemarks, error in
                    guard
                        let placemark = placemarks?.first, error == nil,
                        let administrativeArea = placemark.administrativeArea, //都道府県
                        let locality = placemark.locality, //市区町村
                        let thoroughfare = placemark.thoroughfare, //丁目
                        let subThoroughfare = placemark.subThoroughfare //番地
                            
                        else {
                            self.adrLabel.text = "---"
                            return
                    }
                    
                    self.adrLabel.text = """
                        \(administrativeArea)\(locality)\(thoroughfare)\(subThoroughfare)
                    """
                }
    }
```
(5) GitHubのリンク
[https://github.com/mm0o21/sanphoto](url)

**Comment by appleEmily** [comment]
基本的に、逆ジオコーディングのところについては書いてくれている通り(参考にしてくれた[この記事](https://dev.classmethod.jp/articles/geocoding-use-mapkit/))でよさそう！

方法[1]
①タップされたピンの座標を画面遷移の際に受け渡す。VCの235行目くらいで書いてくれているところね。[値渡し](https://qiita.com/Paulownia_Friends/items/2ba93785cf926e1979b9)
②受け取った座標を使って逆ジオコーディング

方法[2]
①逆ジオコーディング自体をViewControllerの方でやってしまう。
②住所をとれたら、その住所をrealmに保存する。
③editの方では、選ばれたピンの番号に合わせて保存されている住所を表示する。
→この方法だと、画面遷移の時に軽度、緯度を渡さなくて良い

ひとまず言葉で書いてみたので、またわからなかったら聞いてください〜

**Comment by mm0o21** [comment]
二つ目の方法でやってみたのですが、新しいピンを立てようとしたときに "Attempting to create an object of type 'Pin' with an existing primary key value ''."と表示されクラッシュしてしてしまいます。
primary keyが重複？してエラーが出ているのかな？と思ったのですが、解決方法がわからないため教えていただきたいです＞＜

**Reply by appleEmily** [reply]
![image](https://user-images.githubusercontent.com/49332380/192673272-db76cbe9-f6c0-49d2-94f4-f32ffc68dedb.jpeg)![image](https://user-images.githubusercontent.com/49332380/192673277-717e06ca-96e8-4f62-a876-8e8bf4540080.jpeg)

プライマリーキーは、画像のような感じで使ってあげると使えますー！こうしてあげると、エラー自体はまず消えると思うので、そこからやってみてくださいー！
遅くなってごめん🙏

あと、この写真ではもう勝手に書き換えてしまってるんだけど、座標を保存する瞬間に住所も作成して一緒に保存してあげたほうが良いと思います🙆‍♀️

**Reply by mm0o21** [reply]
ありがとうございますー！
